File: dipole_tests/dipole.cpp
#include <vector>
#include <fstream>
#include <iostream>
#include <cmath>
#include "../src/radiation.hpp"
#include <thread>
using namespace std;

// linspace function NOT similar to one in python3
std::vector<double> linspace(double a, double b, size_t m) {
    double h = (b - a) / static_cast<double>(m);
    std::vector<double> xs(m);
	typename std::vector<double>::iterator x;
    double val;
    for (x = xs.begin(), val = a + 0.5 * h; x != xs.end(); ++x, val += h) {
        *x = val;
    }
    return xs;
}

// Note that all normalizations follow those of ../src/radiation.hpp
int main() {

    ofstream vacuum;
    vacuum.open("dipole_data/vacuum.txt"); // distribution d^2I/d\omega d\theta, computed with
                                           // delta_n DO NOT taken into account
    ofstream ref_ind;
    ref_ind.open("dipole_data/ref_ind.txt"); // distribution d^2I/d\omega d\theta, computed with
                                             // delta_n TAKEN into account
	ofstream ch_angle;
	ch_angle.open("dipole_data/ch_angle.txt"); // cherenkov angle as a function of frequency

    ofstream parameters; // contains frequency and angle ranges, etc.
    parameters.open("dipole_data/parameters.txt");

    double gamma_e = 1.0E+8; // particle energy at the begining
	double beta_e = 1 - 0.5 / (gamma_e * gamma_e);
    double b = 0.1; // normalized field
	double mass = 1836; // particle mass ratio to the electron mass
	double omega_L = 0.9; // laser frequency or 1 / a0.
	double x_L = 100; 	// half size of the field region

    // angle varies from 0 to theta_b
	const double theta_right = 0.5 * M_PI;
	const double theta_left  = 0.0;
    size_t m = 27; // number of dots in angle;
    size_t n = 29; // number of dots in frequency ranges
    // angle and omega for the second plot

    vector<double> thetas( linspace( theta_left, theta_right, m) );
    double omega_e = gamma_e / b;
    double omega_left = 0.0;
    double omega_right = omega_e * 0.9;
    vector<double> omegas( linspace(omega_left, omega_right, n) );

    auto I = function<double(double, double, double, double)> (
        [=](double ri, double b, double theta, double omega) {
            auto f = bks_dipole_td(ri, mass, b, x_L, omega_L,gamma_e);
            return omega * b * f(make_tuple(theta, omega)) / (2 * M_PI * r(gamma_e));
        }
    );

    parameters << b           << '\n'
               << m           << '\n'
               << n           << '\n'
               << gamma_e     << '\n'
               << omega_left  << '\n'
               << omega_right << '\n'
               << x_L     	  << '\n'
               << omega_L     << '\n'
               << theta_right << '\n'
               << theta_left  << '\n'
               ;
    for (auto omega: omegas) {
		double ri = vacuum_refractive_index(b, omega);
		ch_angle << acos(1.0 / (ri * beta_e) ) << '\n';
        for (auto theta: thetas) {
            vacuum  << I(1.0, b, theta, omega) << '\n';
            ref_ind << I(ri,  b, theta, omega) << '\n';
        }
    }

    vacuum.close();
    ref_ind.close();
	ch_angle.close();
    parameters.close();

    return 0;
}

File: dipole_tests/dipole.py
import subprocess
import numpy as np
import matplotlib.pyplot as plt
import scipy.special as special
import matplotlib.colors as colors
import matplotlib.cm as mcm
import copy
import scipy.integrate as integrate
from matplotlib.colors import LogNorm

subprocess.run(["make", "dipole"])

params = open("dipole_data/parameters.txt").readlines();
b           = float(params[0]) # magnetic field normalized in accoradance with radiation.hpp
m           =   int(params[1]) # number of dots in theta range
n           =   int(params[2]) # number of dots in omega range
gamma       = float(params[3]) # particle energy
omega_left  = float(params[4])
omega_right = float(params[5])
width_L     = float(params[6])
omega_L     = float(params[7])
theta_right = float(params[8])
theta_left  = float(params[9])
domega = (omega_right - omega_left) / n
chi = gamma * b

omega = np.arange(omega_left, omega_right, domega)
# theta = np.arange(theta_right, theta_left, m)

im1 = np.loadtxt("dipole_data/vacuum.txt") # data with ri = 1
im1 = np.reshape(im1, (n, m))
im2 = np.loadtxt("dipole_data/ref_ind.txt") # data with ri = 1
im2 = np.reshape(im1, (n, m))
ch_angle = np.loadtxt("dipole_data/ch_angle.txt") # cherenkov angle as a function of omega

cmap = 'hot'

fig = plt.figure(figsize = (6, 4.2), constrained_layout = True)
gs = fig.add_gridspec(2, 2)

#================= ax1, left plots ======================
ax1 = fig.add_subplot(gs[0, 0])
vmax = np.max(im1)
im1 = im1 / vmax
im_1 = ax1.imshow( im1, origin = 'lower'
                 , cmap = cmap, aspect = 'auto'
                 , extent = [omega_left/omega_L, omega_right/omega_L, -theta_left, theta_right]
                 , vmin = 0
                 , vmax = 0.1 * np.max(im1)
                 #, norm = LogNorm(vmin = 0.001, vmax = 0.01)
                 )

ax1.set_xlabel("ω/ω_L")
ax1.set_ylabel("θ (rad)")
#================= ax2, right plots ====================
ax2 = fig.add_subplot(gs[0,1])
ax2.plot(omega, ch_angle) 
vmax = np.max(im2)
im2 = im2 / vmax
im_2 = ax2.imshow( im2, origin = 'lower'
                 , cmap = cmap, aspect = 'auto'
                 , extent = [omega_left/omega_L, omega_right/omega_L, -theta_left, theta_right]
                 , vmin = 0
                 , vmax = 0.1 * np.max(im2)
                 )
ax2.set_xlabel("ω/ω_L")
#================= ax3, field float ====================
ax3 = fig.add_subplot(gs[1,0])
sum_im1 = np.sum(im1, axis = 1)
sum_im1 = sum_im1 / np.max(sum_im1)
sum_im2 = np.sum(im2, axis = 1) 
sum_im2 = sum_im2 / np.max(sum_im2)
ax3.plot(omega, sum_im1, "r", label="no-ri")
ax3.plot(omega, sum_im2, "g", label="ri")
ax3.legend(loc="upper left", fontsize = 15)

plt.show()
# plt.savefig('sinus.pdf')
